src/Features/Core/Portable/PullMemberUp/AbstractCodeAndSolutionGenerator.cs

```diff
@@ -0,0 +1,336 @@
+            if (result.Target.TypeKind == TypeKind.Interface)
+            {
+                membersNeedToBeChanged = result.MembersAnalysisResults.
+                Where(analysisResult => analysisResult.ChangeOriginToNonStatic || analysisResult.ChangeOriginToPublic);
+            }
+
+            foreach (var memberAnalysisResult in membersNeedToBeChanged)
+            {
+                var tasks = memberAnalysisResult.Member.DeclaringSyntaxReferences.Select(async @ref => await @ref.GetSyntaxAsync(cancellationToken));
+                var allSyntaxes = await Task.WhenAll(tasks).ConfigureAwait(false);
+                await AddEditorsToDict(allSyntaxes, solutionEditor, solution);
+                SyntaxMap.Add(memberAnalysisResult.Member, allSyntaxes.ToList());
+            }
+
+            if (result.ChangeTargetAbstract)
+            {
+                // When the target is needed to change to abstract, we need the syntax.
+                SyntaxMap.Add(result.Target, new SyntaxNode[] { targetSyntaxNode });
+                await AddEditorsToDict(new SyntaxNode[] { targetSyntaxNode }, solutionEditor, solution);
+            }
+            else
+            {
+                await AddEditorsToDict(new SyntaxNode[] { targetSyntaxNode }, solutionEditor, solution);
+            }
+        }
+
+        private async Task AddEditorsToDict(
+            IEnumerable<SyntaxNode> syntaxNodes,
+            SolutionEditor solutionEditor,
+            Solution solution)
+        {
+            foreach (var syntax in syntaxNodes)
+            {
+                if (!EditorMap.ContainsKey(solution.GetDocumentId(syntax.SyntaxTree)))
+                {
+                    var id = solution.GetDocumentId(syntax.SyntaxTree);
+                    var editor = await solutionEditor.GetDocumentEditorAsync(id).ConfigureAwait(false);
+                    EditorMap.Add(solution.GetDocumentId(syntax.SyntaxTree), editor);
+                }
+            }
+        }
+
+        private ISymbol GetAbstractMemberSymbol(ISymbol symbol)
+        {
+            if (symbol.IsAbstract)
+            {
+                return symbol;
+            }
+
+            var modifier = DeclarationModifiers.From(symbol).WithIsAbstract(true);
+            if (symbol is IMethodSymbol methodSymbol)
+            {
+                return CodeGenerationSymbolFactory.CreateMethodSymbol(methodSymbol, modifiers: modifier);
+            }
+            else if (symbol is IPropertySymbol propertySymbol)
+            {
+                return CodeGenerationSymbolFactory.CreatePropertySymbol(propertySymbol, modifiers: modifier);
+            }
+            else if (symbol is IEventSymbol eventSymbol)
+            {
+                return CodeGenerationSymbolFactory.CreateEventSymbol(eventSymbol, modifiers: modifier);
+            }
+            else
+            {
+                throw new ArgumentException($"{nameof(symbol)} should be method, property, event, indexer");
```

Comment:
thse should not be argument excpetions.  If someone is passing along the wrong kinds, we should fix them.  If this is just trying to assert that we're getting the right type, then just throw `ExceptionUtilities.UnexpectedValue`